src/decoder_pipeline.jl

```diff
@@ -0,0 +1,246 @@
+        if type == sMRZ
+            push!(syndrome_circuit, sMRZ(gate.qubit+s, gate.bit+s))
+        else
+            push!(syndrome_circuit, type(gate.q1, gate.q2+s))
+        end
+    end
+
+    ecirc = encoding_circuit_func(syndrome_circuit)
+    if isnothing(pre_circuit)
+        full_circuit = vcat(pre_circuits, ecirc, errors, noisy_syndrome_circuit)
+    else
+        full_circuit = vcat(ecirc, errors, noisy_syndrome_circuit)
+    end
+
+    frames = PauliFrame(nframes, n+s+k, s+k)
+    pftrajectories(frames, full_circuit)
+    syndromes = pfmeasurements(frames)[:, 1:s]
+    logical_syndromes = pfmeasurements(frames)[:, s+1: s+k]
+
+    for i in 1:nsamples
+        guess = decode(d, syndromes[i])
+
+        result = (O * (guess))[guess_start:guess_stop]
+        
+        if result == logical_syndromes[i]
+            decoded += 1
+        end
+    end
+
+    return (nsamples - decoded) / nsamples
+end
+
+
+"""Type for evaluating the table decoder"""
+struct TableDecoder <: AbstractSyndromeDecoder
+    H
+    faults_matrix
+    n
+    s
+    k
+    lookup_table
+    time 
+end
+
+function TableDecoder(Hx, Hz)
+    c = CSS(Hx, Hz)
+    H = parity_checks(c)
+    s, n = size(H)
+    _, _, r = canonicalize!(Base.copy(H), ranks=true)
+    k = n - r
+    lookup_table, time, _ = @timed create_lookup_table(H)
+    faults_matrix = faults_matrix(H)
+    return TableDecoder(H, n, s, k, faults_matrix, lookup_table, time)
+end
+
+
+"""Type for evaluating the belief propagation decoder"""
+struct BeliefPropDecoder <: AbstractSyndromeDecoder
+    H
+    faults_matrix
+    n
+    s
+    k
+    log_probabs
+    channel_probs
+    numchecks_X
+    b2c_X
+    c2b_X
+    numchecks_Z
+    b2c_Z
+    c2b_Z
+    err
+end
+
+function BeliefPropDecoder(Hx, Hz)
+    c = CSS(Hx, Hz)
+    H = parity_checks(c)
+    s, n = size(H)
+    _, _, r = canonicalize!(Base.copy(H), ranks=true)
+    k = n - r
+    println(typeof(H))
+    faults_matrix = faults_matrix(H)
+    log_probabs = zeros(n)
+    channel_probs = fill(p_init, n)
+
+    numchecks_X = size(Cx)[1]
+    b2c_X = zeros(numchecks_X, n)
+    c2b_X = zeros(numchecks_X, n)
+
+    numchecks_Z = size(Cz)[1]
+    b2c_Z = zeros(numchecks_Z, n)
+    c2b_Z = zeros(numchecks_Z, n)
+    err = zeros(n)
+    return BeliefPropDecoder(H, faults_matrix, n, s, k, log_probabs, channel_probs, numchecks_X, b2c_X, c2b_X, numchecks_Z, b2c_Z, c2b_Z, err)
+end
+
+
+"""Decoding function for the table decoder"""
+function decode(d::TableDecoder, syndrome_sample)
+    return get(d.lookup_table, syndrome_sample, nothing)
+end
+
+"""Decoding function for the belief propagation decoder"""
+function decode(d::BeliefPropDecoder, syndrome_sample)
+    row_x = syndrome_sample[1:d.numchecks_X]
+    row_z = syndrome_sample[d.numchecks_X+1:d.numchecks_X+d.numchecks_Z]
+
+    KguessX, success = syndrome_decode(sparse(d.Cx), sparse(d.Cx'), d.row_x, d.max_iters, d.channel_probs, d.b2c_X, d.c2b_X, d.log_probabs, Base.copy(d.err))
```

Comment:
If these sparse matrices are necessary, these should be created in advance, not each time a decoder is called. These can be, for instance, additional properties of the struct.